src/main/java/io/jenkins/plugins/github/checks/GitHubChecksClient.java

```diff
@@ -47,7 +47,8 @@ public void createCheckRun(ChecksContext context, ChecksDetails details) {
                 new GitHubBuilder().withAppInstallationToken(token).build()
                         .getRepository(repository).createCheckRun(details.getName(), headSha)
                         .withStartedAt(new Date())
-                        .withDetailsURL(details.getDetailsURL())
+                        .withDetailsURL(details.getDetailsURL().isPresent() ?
```

Comment:
use `orElse`
```suggestion
                        .withDetailsURL(details.getDetailsURL().orElse(context.getRun().getUrl())
```

Reply:
This depends a little bit on the answer of https://github.com/XiongKezhi/checks-api-plugin/pull/14#discussion_r436266916.

If we really need the `Context` type, then this is the way to go. I would suggest one minor thing ([tell don't ask](https://martinfowler.com/bliki/TellDontAsk.html)): 

```suggestion
                        .withDetailsURL(details.getDetailsURL().orElse(context.getDetailsUrl()))
```

If we can embed the `Context` information into the `Details` it would simply resolve to

```suggestion
                        .withDetailsURL(details.getDetailsURL())
```

and the `Details` instance handles everything for us

Reply:
Embed `Context` in `Details` is a good idea, but one problem is that we then may need to add `run` as another parameter for the constructor, is that ok?

Reply:
For the immediate publishing we need to do something like (see https://github.com/XiongKezhi/checks-api-plugin/pull/14/files#r436286430)

```
Run run = ...;
// find a Publisher for our SCM provider
ChecksPublisher publisher = ChecksPublisherFactory.find(run); 

// or if we need the SCM as well
// ChecksPublisher publisher = ChecksPublisherFactory.find(run, scm);

DetailsBuilder builder = new DetailsBuilder("CheckStyle", run);

// Or the publisher sets the run in a factory method
// or DetailsBuilder builder = publisher.createDetailsBuilder("Check style")

// Or the publisher sets the run after the publish method
    
builder.withStatus(PROGRESS);

publisher.publishChecks(run, builder.build());
```

So in this case the run must be available anyway. We can retrieve from the publisher or we can add it as parameter. I'm not sure how we handle the second case (queued and therefore no run available yet).

Reply:
But what about the `name` duplication problem? Before I was thinking `ChecksPublisher` to be extended by consumers and they will implement the `getName()`, so we know all the consumer checks names, but now, the control is in the hand of the users, how can we solve that?

Reply:
I don't understand why we do not need a factory method to obtain the correct publisher? Is there another way to get the publisher? The publisher is required to invoke the `publishChecks` method.  How do you obtain it currently?